### ISSUE TITLE
Speed up dependency checks for large numbers of outputs

### ISSUE THREAD

**Issue by trwhitcomb** [cylc/cylc-flow#3255]
We ran into a case where suites were taking a very long time to validate.  This was covered in #1776 and follow on tickets, but didn't quite solve the problem.  We tracked the behavior down to an O(N**2) operation in the number of task outputs.

Steps to reproduce: Have a very large number of task outputs (~1000)? in the graph that will be loaded.

Our solution was to replace a list with a set in the taskdef initialization that allowed checking for membership to be done in O(1) time instead of O(N) time (PR to follow)

**Comment by hjoliver** [comment]
Hi @trwhitcomb - that's great news, well done!  Looking forward to the PR...

**Comment by trwhitcomb** [comment]
A few notes: 

1. (I found this while backporting this to our 6.11.4 install (I know)) - if the objects placed into a set are not hashable (as they were back then, with taskdef outputs as custom Python objects), even the set wasn't enough to avoid the O(N) test for membership.  However, this went away once I defined a custom `__hash__` function for the taskdef output object.  In v7, these things are just tuples (IIRC) so no special treatment is necessary beyond the changes in this PR.

2. The symptoms that sent us looking for this were exactly the behavior outlined in #2306 - the first complaints is that suites were "unresponsive on startup/restart" and would not respond to any commands.  We were able to isolate this to just the validate command taking forever, which led to identification of the problem.  Hat tip to @alpacaxander for his debugging work.

3. This addresses the underlying root cause of the behavior in #2306, I would vote not to close that other ticket as hitting these large, expensive-to-validate suites shouldn't cause the suite server to go numb to the outside world.
